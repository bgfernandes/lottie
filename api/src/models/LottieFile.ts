import { JSONSchema7 } from 'json-schema'
import BaseModel from './BaseModel'

export default class LottieFile extends BaseModel {
  id: string | undefined
  slug: string | undefined
  url: string | undefined

  static get tableName():string {
    return 'lottie_files'
  }

  static get jsonSchema():JSONSchema7 {
    return {
      type: 'object',

      properties: {
        id: { type: 'string' },
        slug: { type: 'string', minLength: 21, maxLength: 21 },
        url: { type: 'string' },
        created_at: { type: 'string' },
        updated_at: { type: 'string' }
      },

      required: ['slug', 'url']
    }
  }
}
