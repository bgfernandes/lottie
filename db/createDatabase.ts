/*
  Helper script file to create the DB
*/

import knex from 'knex'
import dotenv from 'dotenv'
import path from 'path'

if (require.main === module) {
  dotenv.config({
    path: path.resolve(__dirname, '..', '.env.local')
  })

  createDatabase()
}

export default function createDatabase():Promise<void> {
  const knexInstance = knex({
    client: 'pg',
    connection: {
      host : process.env.DB_HOST,
      user : process.env.DB_USER,
      password : process.env.DB_PASS
    }
  })

  return knexInstance.raw(`CREATE DATABASE ${process.env.DB_DATABASE};`)
    .then(() => {
      console.log(`Created Database ${process.env.DB_DATABASE}`)
    })
    .catch((err) => {
      if (err.message.match(/^CREATE DATABASE .*; - database ".*" already exists$/)) {
        console.log(`Database ${process.env.DB_DATABASE} already exists.`)
      } else {
        console.error(`Error creating database ${process.env.DB_DATABASE}`)
        console.error(err)
      }
    })
    .finally(() => {
      knexInstance.destroy()
    })
}
