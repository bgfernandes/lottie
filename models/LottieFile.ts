import { JSONSchema7 } from 'json-schema'
import BaseModel from './BaseModel'

export default class LottieFile extends BaseModel {
  id: string | undefined;

  static get tableName():string {
    return 'lottie_files'
  }

  static get jsonSchema():JSONSchema7 {
    return {
      type: 'object',

      properties: {
        id: { type: 'string' },
        created_at: { type: 'string' },
        updated_at: { type: 'string' }
      }
    }
  }
}
