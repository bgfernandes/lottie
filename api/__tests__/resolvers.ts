/*
  Integration tests for the Apollo Server, uses the DB
*/

import { ApolloServer } from 'apollo-server-express'
import knexCleaner from 'knex-cleaner'
import { gql } from 'apollo-server-core'
import { Model } from 'objection'
import moment from 'moment'
import createApolloServer from '../src/apollo-server'
import createDatabase from '../db/createDatabase'
import LottieFile from '../src/models/LottieFile'
import initializeDb from '../src/config/initializeDb'

let server: ApolloServer

const HELLO_MESSAGE_QUERY = gql`
  query HelloMessage {
    hello {
      message
    }
  }
`

describe('Resolvers', () => {
  beforeAll(async () => {
    // create test database if not already created, run migrations, and setup objection
    await createDatabase()
    initializeDb()
    await Model.knex().migrate.latest({
      directory: './db/migrations'
    })

    server = createApolloServer()
  })

  afterAll(async () => {
    await Model.knex().destroy()
  })

  beforeEach(async () => {
    await knexCleaner
      .clean(Model.knex(), { ignoreTables: ['knex_migrations', 'knex_migrations_lock'] })
  })

  it('runs the home query successfully', async () => {
    const res = await server.executeOperation({ query: HELLO_MESSAGE_QUERY })
    expect(res).toMatchSnapshot()
  })

  describe('lottieFiles query', () => {
    it('runs the lottieFiles query successfuly and returns data', async () => {
      await LottieFile.query().insert({
        id: '1',
        slug: 'W6HDcJoY946T8gGYLiiAJ',
        url: 'http://some_url.json',
        createdAt: moment('2021-09-01 00:00:00Z').toISOString(),
        updatedAt: moment('2021-09-02 00:00:00Z').toISOString()
      })
      await LottieFile.query().insert({
        id: '2',
        slug: 'W6HDcJoY946T8gGYLiiA2',
        url: 'http://some_url2.json',
        createdAt: moment('2021-09-03 00:00:00Z').toISOString(),
        updatedAt: moment('2021-09-04 00:00:00Z').toISOString()
      })

      const LOTTIE_FILES_QUERY = gql`
        query {
          lottieFiles {
            id,
            createdAt,
            updatedAt
          }
        }
      `
      const res = await server.executeOperation({ query:  LOTTIE_FILES_QUERY})
      expect(res).toMatchSnapshot()
    })
  })
})
