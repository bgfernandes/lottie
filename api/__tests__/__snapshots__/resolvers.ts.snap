// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers lottieFile query returns empty when the lottie is not found 1`] = `
Object {
  "data": Object {
    "lottieFile": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolvers lottieFile query returns the lottie successfully 1`] = `
Object {
  "data": Object {
    "lottieFile": Object {
      "createdAt": "1630454400000",
      "id": "1",
      "slug": "W6HDcJoY946T8gGYLiiAJ",
      "updatedAt": "1630540800000",
      "url": "http://some_url.json",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolvers lottieFiles query runs the lottieFiles query successfuly and returns data 1`] = `
Object {
  "data": Object {
    "lottieFiles": Array [
      Object {
        "createdAt": "1630454400000",
        "id": "1",
        "updatedAt": "1630540800000",
      },
      Object {
        "createdAt": "1630627200000",
        "id": "2",
        "updatedAt": "1630713600000",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolvers runs the home query successfully 1`] = `
Object {
  "data": Object {
    "hello": Object {
      "message": "Hello from Apollo.",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
